processor/processor.go

```diff
@@ -1372,6 +1372,107 @@ func getDiffMetrics(inPU, pu string, inCountMetadataMap map[string]MetricMetadat
 	return diffMetrics
 }
 
+func generateFilteredAndAddedMetrics(inPU, pu string, messageIds []string, messageIdKeyMap map[string]string, inCountMetadataMap map[string]MetricMetadata, outputEvents []transformer.TransformerEvent, failedEvents []transformer.TransformerResponse) ([]*types.PUReportedMetric, []*types.PUReportedMetric) {
+	successMessageIdMap := make(map[string]int)
+	failedMessageIdMap := make(map[string]int)
+	for _, event := range outputEvents {
+		// handle transformBatch user transformation
+		// consider it as success if messageId is present in Metadata.MessageIDs of output event
+		// but do not increment to not count it as newly added event
+		if event.Metadata.MessageID == "" {
+			for _, messageId := range event.Metadata.MessageIDs {
+				if _, ok := successMessageIdMap[messageId]; !ok {
+					successMessageIdMap[messageId] = 0
+				}
+			}
+			continue
+		}
+
+		if _, ok := successMessageIdMap[event.Metadata.MessageID]; !ok {
+			successMessageIdMap[event.Metadata.MessageID] = 0
+		}
+		successMessageIdMap[event.Metadata.MessageID] += 1
+	}
+	for _, event := range failedEvents {
+		// handle transformBatch user transformation
+		// consider it as failed if messageId is present in Metadata.MessageIDs of failed event
+		if event.Metadata.MessageID == "" {
+			for _, messageId := range event.Metadata.MessageIDs {
+				if _, ok := failedMessageIdMap[messageId]; !ok {
+					failedMessageIdMap[messageId] = 0
+				}
+				failedMessageIdMap[messageId] += 1
+			}
+			continue
+		}
+
+		if _, ok := failedMessageIdMap[event.Metadata.MessageID]; !ok {
+			failedMessageIdMap[event.Metadata.MessageID] = 0
+		}
+		failedMessageIdMap[event.Metadata.MessageID] += 1
+	}
+
+	filteredMetricsMap := make(map[string]*types.PUReportedMetric)
+	addedMetricsMap := make(map[string]*types.PUReportedMetric)
+	for _, messageId := range messageIds {
+		outCount, successOk := successMessageIdMap[messageId]
+		_, failedOk := failedMessageIdMap[messageId]
+		noMessageIdInOutput := !successOk && !failedOk
+		multipleMessageIdInOutput := successOk && outCount > 1
+
+		if noMessageIdInOutput || multipleMessageIdInOutput {
+			key := messageIdKeyMap[messageId]
+			metricMetadata := inCountMetadataMap[key]
+			var eventName, eventType string
+			splitKey := strings.Split(key, MetricKeyDelimiter)
```

Comment:
it is about time to prepare a typed reportingKey, so that we don't need to repeat ourselves and perform complex parsing calculations, e.g.

```go
type metricKey struct {
	sourceID       string
	destinationID  string
	sourceJobRunID string
	eventName      string
	eventType      string
}
```

Reply:
done